result.nim

```diff
@@ -0,0 +1,352 @@
+    ## More C++: https://github.com/ned14/outcome
+    ##
+    ## Swift is interesting in that it uses a non-exception implementation but
+    ## calls errors exceptions and has lots of syntactic sugar to make them feel
+    ## that way by implicitly passing them up the call chain - with a mandatory
+    ## annotation that function may throw:
+    ## https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/ErrorHandling.html
+    ##
+    ## # Other implemenations in nim
+    ##
+    ## There are other implementations in nim that you might prefer:
+    ## * Either from nimfp: https://github.com/vegansk/nimfp/blob/master/src/fp/either.nim
+    ## * result_type: https://github.com/kapralos/result_type/
+    ##
+    ## # Implementation notes
+    ##
+    ## This implementation is mostly based on the one in rust. Compared to it,
+    ## there are a few differences - if know of creative ways to improve things,
+    ## I'm all ears.
+    ##
+    ## * Rust has the enum variants which lend themselves to nice construction
+    ##   where the full Result type isn't needed: `Err("some error")` doesn't
+    ##   need to know value type - maybe some creative converter or something
+    ##   can deal with this?
+    ## * Nim templates allow us to fail fast without extra effort, meaning the
+    ##   other side of `and`/`or` isn't evaluated unless necessary - nice!
+    ## * In Nim, we have exceptions - when using this library, we'll raise the
+    ##   standard crop of Nim errors when trying to access the error of a value
+    ##   and vice versa - this fits better with Nim but costs space and
+    ##   performance - need to think about this - rust will simply panic, and
+    ##   everyone seems more or less happy with that..
+    ## * Rust uses From traits to deal with result translation as the result
+    ##   travels up the call stack - needs more tinkering - some implicit
+    ##   conversions would be nice here
+
+    case isOk*: bool
+    of false:
+      error*: E
+    of true:
+      value*: T
+
+  # TODO cannot raise generic error, so cannot include error value :(
+  # https://github.com/nim-lang/Nim/issues/7845
+  ResultError = object of Exception
+
+# TODO check that r is a Result
+proc ok*(r: typedesc, v: auto): r =
+  ## Initialize a result with a success and value
+  ## Example: `Result[int].ok(42)`
+  r(isOk: true, value: v)
+proc ok*[T, E](self: var Result[T, E], v: auto) =
+  ## Set the result to success and update value
+  ## Example: `result.ok(42)`
+  self = Result[T, E].ok(v)
+
+proc err*(r: typedesc, e: auto): r =
+  ## Initialize the result to an error
+  ## Example: `Result[int].err("uh-oh")`
+  r(isOk: false, error: e)
+
+proc err*[T, E](self: var Result[T, E], v: auto) =
+  ## Set the result as an error
+  ## Example: `result.err("uh-oh")`
+  self = Result[T, E].err(v)
+
+template isErr*[T, E](self: Result[T, E]): bool = not self.isOk
```

Comment:
I know this style hasn't gained many followers in the Nim community yet, but I usually recommend to use implicit generics:

``` nim
template isErr*(self: Result): bool = not self.isOk
```

I think the shorter code is more visually pleasing and it's more resilient to re-factoring (You can make changes to the generic parameters without touching all the procs in the module).

Reply:
ah yes, this is nicer